### PR TITLE
[#3] Create data types

### LINKED ISSUE #3: Implement all data types, described in README
Data types only. No functions. The idea is to implement data model in Haskell and then change it if needed.

Also in real world you don't work with single TODO item. You have some collection of TODO items. So you should implement type like `TodoList` which store your TODO items.

Don't think about history of changes for now. History of changes deserves its own design.

### PR DESCRIPTION
**Pull request by StasyaGera**
Resolves #3 

This code lacks `instance` specifications, but the issue was only about types and `instance` requires functions (am I right?).

Creating `TodoList` makes no sense for me so far, could you explain why do we need a special type for it?

### CODE REVIEW

**Comment by ChShersh** [comment]
> This code lacks instance specifications, but the issue was only about types and instance requires functions (am I right?).

Yeah, no type classes needed. Also, for now all required instances can be derived. I don't see any instances you need to write manually...

> Creating TodoList makes no sense for me so far, could you explain why do we need a special type for it?

Well, users won't have single `Todo` item in their life. Usually you work with multiple todo items, like:

1. Get shopping.
2. Learn Japanese.

So you at least need list. But `[Todo]` is not actually a best decision, but ok for now :ok_hand: 

I think I know what you mean here. Since `Todo` item stores list of subtasks internally, you don't need to have list of top-level TODO items. You will have only single root `TODO` item. All TODO items will be added as subtasks for this root TODO item. I can't say this solution is good. Also, I can't say, this solution is bad :frowning: Always having some root TODO item looks weird, because it means that there's special TODO. And this not represented in types. And why data structure which stores top level todo items is also a TODO? Though, not needing extra data type for storing top-level todo items is also good. But for now I tend to vote for `type Todos = [Todo]`.

**Comment by ChShersh** [comment]
Regarding TODO root and top level list of TODO items: I've created issue #6 

So, please, don't do anything related to `TodoList` in this PR. This will be done in separate PR under separate issue. But you can read my thoughts and participate in discussion :slightly_smiling_face:

File: src/Lib.hs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+module Lib where
+
+import           Data.List.NonEmpty
+
+data State = TODO | DONE | FAIL deriving (Eq, Show)
+
+newtype Title = Title (NonEmpty Char) deriving (Eq, Show)
+newtype Priority = Priority Int deriving (Eq, Show)
+newtype Tag = Tag String deriving (Eq, Show)
+newtype Tags = Tags (NonEmpty Tag) deriving (Eq, Show)
```
[the remark this answers is not in the document]
Comment (StasyaGera) [reply src/Lib.hs]: So am I supposed to change these to just `type`s? (I see four lines associated with your comment, but you used "this", did you mean only one of them?)

And I also have a question about `newtype`: does the newly defined type inherit type class specifications?
I mean, should you explicitly write `instance Ord Priority`, for example, and specify functions or will it work like `Int` does?

[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Lib.hs]: > [upd:] I reckon you meant changing only type Tags.

Yes, I mean only `Tags`. Other `newtype`s are good! Sorry for being not clear enough.

> I mean, should you explicitly write instance Ord Priority, for example, and specify functions or will it work like Int does?

You use `deriving`. For `newtype` you can derive any type class for which unwrapped data type has instance. For example, for `Priority` you can even put `Num` into deriving! I don't now whether  it's a good idea to derive `Num` for `Priority` but it's possible. And you can derive `Ord` just as you derived `Eq` and `Show`.

And you should enable `GeneralizedNewtypeDeriving` extension by default in `.cabal` file.

File: src/Todo/Model.hs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module Todo.Model where
+
+import           Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.Ratio         as R
+import qualified Numeric.Natural    as N
+
+data State = TODO | DONE | FAIL deriving (Eq, Show)
+
+newtype Title = Title (NonEmpty Char) deriving (Eq, Show)
+newtype Priority = Priority (R.Ratio N.Natural) deriving (Eq, Show)
```
Comment (ChShersh) [inline src/Todo/Model.hs]: I don't see comment regarding order of `Priority`...

File: src/Todo/Model.hs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+module Todo.Model where
```
Comment (ChShersh) [inline src/Todo/Model.hs]: Could you please also specify explicit export list? This is a good design for projects. And it helps in future. To export data type and all its fields and constructors you need to write something like that:

```haskell
module Todo.Model where
       ( Todo (..)
       , State (..)
       , Title (..)
       -- the rest
       ) where
```

File: src/Todo/Model.hs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+newtype Tag = Tag (NonEmpty Char) deriving (Eq, Show)
+newtype Deadline = Deadline Int deriving (Eq, Show)
+type Tags = NonEmpty Tag
+
+data Todo = Todo
+    { todoState       :: State
+    , todoTitle       :: Title
+    , todoPriority    :: Priority
+    , todoTags        :: Tags
+    , todoSubtasks    :: [Todo]
```
Comment (ChShersh) [inline src/Todo/Model.hs]: Can be just `Todos`. But if you don't like `Todos` type alias you can safely remove it.

File: todo-cli.cabal [hunk 5]
```diff
@@ -0,0 +1,54 @@
+-- This file has been generated from package.yaml by hpack version 0.20.0.
+--
+-- see: https://github.com/sol/hpack
+--
+-- hash: 592e31a8bba4a31e2540e7aea82b3da47430f4c1a8a18f898de13cd3be53b9a9
```
Comment (ChShersh) [inline todo-cli.cabal]: Okay, I assume you don't have `package.yaml` anymore in your project :) So let's prettify this `.cabal` file. And start with removing this comment and empty line before `name`.

File: todo-cli.cabal [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+homepage:       https://github.com/StasyaGera/todo-cli#readme
+bug-reports:    https://github.com/StasyaGera/todo-cli/issues
+author:         Stanislava Gera
+maintainer:     stasya.gera@yandex.ru
+copyright:      2018 Stanislava Gera
+license:        BSD3
+license-file:   LICENSE
+build-type:     Simple
+cabal-version:  >= 1.10
+
```
Comment (ChShersh) [inline todo-cli.cabal]: Empty line here is redundant.

File: todo-cli.cabal [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+extra-source-files:
+    ChangeLog.md
+    README.md
+
+source-repository head
+  type: git
+  location: https://github.com/StasyaGera/todo-cli
+
+library
+  hs-source-dirs:
```
Comment (ChShersh) [inline todo-cli.cabal]: Could you please format all fields here just like in this project?

* https://github.com/serokell/importify/blob/master/importify.cabal

Aligning field names to the left and field values to the right helps to observe values easier.

File: todo-cli.cabal [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      base >= 4.9 && < 5
+  ghc-options:
+      -Wall
+  exposed-modules:
+      Todo.Model
+  other-modules:
+      Paths_todo_cli
+  default-language: 
+      Haskell2010
+  default-extensions: RecordWildCards
```
Comment (ChShersh) [inline todo-cli.cabal]: As I mentioned, you should also add `GeneralizedNewtypeDeriving` here

File: todo-cli.cabal [hunk 9]
```diff
@@ -1,54 +1,38 @@
...
-      Haskell2010
-  default-extensions: RecordWildCards
+  hs-source-dirs:       src
+  build-depends:        base >= 4.9 && < 5
+  ghc-options:          -Wall
+  exposed-modules:      Todo.Model
+  other-modules:        Paths_todo_cli
+  default-language:     Haskell2010
+  default-extensions:   RecordWildCards
+                      , GeneralizedNewtypeDeriving
```
Comment (ChShersh) [inline todo-cli.cabal]: `,` is not needed in `default-extensions`. But this is not important for now, you can remove it when you add another extension here.

**Comment by ChShersh** [comment]
You can now press `Squash and merge` button :sunglasses:

**Pull request merged**